lib/Mojo/Redis/Protocol.pm

```diff
@@ -0,0 +1,240 @@
+package Mojo::Redis::Protocol;
+use Mojo::Base -base;
+
+use Carp qw(confess croak);
+use Encode 'find_encoding';
+
+use constant DEBUG => $ENV{MOJO_REDIS_DEBUG} || 0;
+
+our %ENCODING;    # Internal usage
+
+has on_message => undef;
+
+sub encode {
+  my ($self, @stack) = @_;
+
+  my $str = '';
+  while (@stack) {
+    my $message = shift @stack;
+    my $data    = $message->{data};
+    my $type    = $message->{type};
+
+    if ($type eq '+' or $type eq '-' or $type eq ':') {
+      $str .= "$type$data\r\n";
+    }
+    elsif ($type eq '$') {
+      $str .= defined $data ? '$' . length($data) . "\r\n$data\r\n" : "\$-1\r\n";
+    }
+    elsif ($type eq '*') {
+      if (defined $data) {
+        $str .= '*' . scalar(@$data) . "\r\n";
+        unshift @stack, @$data;
+      }
+      else {
+        $str .= "*-1\r\n";
+      }
+    }
+    else {
+      confess "[Mojo::Redis::Protocol] Unknown message type $type";
+    }
+  }
+
+  return $self->{encoder} ? $self->{encoder}->encode($str, 0) : $str;
```

Comment:
The encoding should be applied to the data component, not the composed message.

Reply:
Can you explain what the practical difference is?

Reply:
$message->{data} is what you want to encode and decode, not the structure of the protocol which always needs to be the same bytes.

Reply:
I understand that part, I just don’t see in which case it matters.

Reply:
With any encoding which has a different byte representation of the protocol elements such as $, \r and \n. Only the bulk string content can be encoded really, and this may change its length, everything else needs to be ascii bytes that redis expects.

Reply:
This should be fixed now.